### PR TITLE
Fix RGB 565 to 888

### PR DESCRIPTION
**Pull request by paulpv**
The initial commit is **only** for 16-bit 565 to 32-bit 888, and is mostly just to get the conversation going before marking this Ready For Review.

Rather than just link to a post in the forum (https://forum.littlevgl.com/t/pre-pr-proposal-improve-565-to-888-color-conversions-and-perhaps-others/836), I will copy my post here so that this PR is more self-contained.

I’ve talked w/ @kisvegabor  and @embeddedt a bit about this in a semi-unrelated PR (#1211).

I believe that LittlevGL’s current 565 to 888 conversion code is flawed (which opens up the possibility of others to/from)
```
    lv_misc/lv_color.h:
    static inline uint32_t lv_color_to32(lv_color_t color)
    {
        ...
        #elif LV_COLOR_DEPTH == 16
        lv_color32_t ret;
        ret.ch.red = color.ch.red * 8; /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
        ret.ch.green = color.ch.green * 4; /*(2^8 - 1)/(2^6 - 1) = 255/63 = 4*/
        ret.ch.blue = color.ch.blue * 8; /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
        ret.ch.alpha = 0xFF;
        return ret.full;
        ...
    }
```
I think the above code should cleanly convert 565 LV_COLOR_RED of 0xF800 to 888 of 0xFF0000, but it doesn’t.
I understand the important performance rationale for 255/31=8 and not using an actual floating point of 8.22580645 and the loss that integer division introduces; minimizing this loss is actually one of my points for suggesting an improvement.
The 5 bit red channel of 0xF800 is 2^5-1 or 31 or 0x1F.
Per the above code logic, 31 * 8 is 248, or 0xF8.
Thus, the above code converts 565 0xF800 to 888 of 0xF80000 instead of 0xFF0000.
I understand that 565 to 888 conversion is “lossy/imperfect”, but when mapping a number from one scale to another, I believe at least the min and max values should convert without loss.
Thus, I think that the above code needs to change.

There is a mystery in the numbers of this commonly used more precise RGB 565 to 888 conversion code:
SO Q&A: [How does one convert 16-bit RGB565 to 24-bit RGB888](https://stackoverflow.com/a/9069480/252308)?
```
    R8 = ( R5 * 527 + 23 ) >> 6;
    G8 = ( G6 * 259 + 33 ) >> 6;
    B8 = ( B5 * 527 + 23 ) >> 6;
```
The discussion in the above SO answer suggests 527 may have been incorrectly rounded up instead of down and should be 526 (which if true would also change the 23 addition.
Searching the internets gave me no clue as to the origin of these values.
I have created a spreadsheet exploring how these values were derived.
https://docs.google.com/spreadsheets/d/1PppX8FJpddauAPasHwlNgIPGIuPGPNvRbhilIQ8w-7g/edit#gid=0
It seems like nearly any listed multiplier and adder can be used, but it is still a mystery to me why those specific 6 bit shift values were chosen over any others.

@kisvegabor and @embeddedt agree that the colors seem off.

The simple 565 to 888 fix is to change the equation from the above OP’s simple * 8 and * 4 to either Apple’s version, a common/popular version on the interwebs, or one that I derived.
Apple:
```
    /*
     * Per https://developer.apple.com/documentation/accelerate/1533159-vimageconvert_rgb565toargb8888
     * Apple must have an ingenious reason to divide an integer by 63 or prime 31!
     */
    ret.ch.red   = ( color.ch.red * 255 + 15 ) / 31;
    ret.ch.green = ( color.ch.green * 255 + 31 ) / 63;
    ret.ch.blue  = ( color.ch.blue * 255 + 15 ) / 31;
```
Popular:
```
    /**
     * Per https://stackoverflow.com/a/9069480/252308
     * These values can be searched for and seen to be used in many places.
     */
    ret.ch.red   = ( color.ch.red * 527 + 23 ) >> 6;
    ret.ch.green = ( color.ch.green * 259 + 33 ) >> 6;
    ret.ch.blue  = ( color.ch.blue * 527 + 23 ) >> 6;
```
Mine:
```
    /**
     * Per https://docs.google.com/spreadsheets/d/1PppX8FJpddauAPasHwlNgIPGIuPGPNvRbhilIQ8w-7g/edit#gid=0
     * Truly any of the listed multipliers and adders would work
     */
    ret.ch.red   = ( color.ch.red * 526 + 14 ) >> 6;
    ret.ch.green = ( color.ch.green * 259 + 3 ) >> 6;
    ret.ch.blue  = ( color.ch.blue * 526 + 14 ) >> 6;
```

Which do you recommend?
In all modesty, I understand my numbers and not the others, so I recommend mine.
Truly I think the following are the most precise best, but they seem radically different than anything else I have seen out there, so I’m a bit wary to use them:
```
    ret.ch.red   = ( color.ch.red * 33 - 3 ) >> 2;
    ret.ch.green = ( color.ch.green * 4 + 3 );
    ret.ch.blue  = ( color.ch.blue * 33 - 3 ) >> 2;
```

In the end I went with these, but I can use any of them or any others that may be better...as long as they are accurate.

### CODE REVIEW

**Comment by paulpv** [comment]
In addition to approving if the pushed code for 565 to 888 seems fine, the other conversation that needs to be had is if there are any other errors in the **other** bit conversion code (to/from 1/16/32), and how to best fix them.
I'd prefer this PR be exhaustive to fix any and all bit conversion problems.
Naturally [to me] this segues toward having some sort of unit tests.
Or course this opens a can of worms.

Should this PR only fix 565, or all bit conversions?
Should this PR include some/any unit tests (and all the baggage that may include), or should we just keep on keepin on?

**Comment by embeddedt** [comment]
> Should this PR include some/any unit tests (and all the baggage that may include), or should we just keep on keepin on?

I think we should keep unit tests out of this PR, because we haven't really decided how those are going to work yet. That's not saying that I don't want unit testing; it's just that I don't think we should clutter this PR too much with that.

For this, I think someone should just pick some colors that are expected to work and manually try converting them.

> I can use any of them or any others that may be better...as long as they are accurate.

We can't use the StackOverflow one, because by default all SO code is licensed under CC-BY-SA (or a similar CC license) and that can't be included in a codebase that advertises itself as MIT (as far as I know, but I'm not a lawyer).

I'm wary of using the Apple one because it has no license provided.

I would rather use the one you derived or the "most precise" one, since those two appear to have originated from you.

**Comment by kisvegabor** [comment]
I agree with @embeddedt about both the uni test and licenses. I've found this about [SO license](https://stackoverflow.blog/2009/06/25/attribution-required/)

@paulpv I've tried the new values and worked great for me! :+1:

**Comment by paulpv** [comment]
In the interest of small changes, should we merge this PR as is, or investigate the other conversions too?

**Comment by embeddedt** [comment]
Merging as-is should be fine, although, to be honest, it really doesn't matter.

**Comment by kisvegabor** [comment]
@paulpv 
I've noticed that the current algorithm doesn't work if `red` or `blue` is `0`.
```c
( color.ch.red * 33 - 3 ) >> 2; // (0 * 33 - 3) = -3 -> 252 (undeflow) 
```

**Comment by embeddedt** [comment]
Cross-posting this - someone else discovered what I think is the same issue with the HSV conversion:

https://forum.littlevgl.com/t/bug-when-transform-color-by-lv-color-rgb-to-hsv-and-transform-backward-by-lv-color-hsv-to-rgb-it-cant-display-the-same-lv-color-t-color/924?u=embeddedt

**Comment by paulpv** [comment]
I haven't done the actual math yet, but my first thought is that the rounding error w/ integer math can be taken advantage of to cause this to code to be just:
```
    ret.ch.red   = ( color.ch.red * 33 ) >> 2; 
    ret.ch.green = ( color.ch.green * 4 );
    ret.ch.blue  = ( color.ch.blue * 33 ) >> 2;
```
I will toy w/ integer based math on my spreadsheet to dial this in...

**Comment by paulpv** [comment]
(0 * 33) >> 2 = 0
(31 * 33) >> 2 = 255

**Comment by paulpv** [comment]
Please give this a glance and see what you think.
I suspect that as long as the +/- factor is less than the integer rounding error, then the +/- can be safely ignored.
I will also update the comment with a bit more clarity.

Regarding HSV, that is a separate issue that I will address in separate PR that I am still working on.

**Comment by paulpv** [comment]
Problem for green:
( 0 * 4 ) = 0
( 63 * 4 ) = 252
I'll tweak...

**Comment by paulpv** [comment]
Working the spreadsheet, the trick is to find the first row that has a + adder (to prevent underflow) and a `min` result of `0` and `max` result of `255`.
I've settled on the following conversions, and they seem to pass all test bounds:
```
ret.ch.red   = ( color.ch.red * 8423 + 7 ) >> 10; 
ret.ch.green = ( color.ch.green * 259 + 3 ) >> 6;
ret.ch.blue  = ( color.ch.blue * 8423 + 7 ) >> 10;
```

**Comment by paulpv** [comment]
Ready for final final review and hopefully merge

**Comment by paulpv** [comment]
FYI, as mentioned in the OP, it would be best to go through a similar experiment with all other color bit conversions and find out if any of them are off too.

**Comment by embeddedt** [comment]
@TridentTD reported that their issue hasn't been solved yet:

https://forum.littlevgl.com/t/bug-when-transform-color-by-lv-color-rgb-to-hsv-and-transform-backward-by-lv-color-hsv-to-rgb-it-cant-display-the-same-lv-color-t-color/924/4?u=embeddedt

**Comment by paulpv** [comment]
@embeddedt as expected. I haven't fixed the hsv problem yet. That is on a separate branch that is still in progress and isn't PRed yet.

**Comment by TridentTD** [comment]
The new algorithm from @paulpv 
for transforming color16bits to color32bits 
and transform backward by `lv_color_make(...)`.

I tested by the following code.
```c
  lv_color_t    color16;
  lv_color32_t  color32;
  lv_color16_t  color16_2;
  for(uint32_t c = 0; c < 0xFFFF; c++) {
    color16.full = c;
    color32.full = lv_color_to32(color16);
    color16_2    = lv_color_make(color32.ch.red,color32.ch.green, color32.ch.blue);
    if(color16.full != color16_2.full){
      Serial.printf("Color16 : Before 0x%04x --> After 0x%04x\n", color16.full, color16_2.full);
    }
    if( c % 1000 == 0) Serial.printf("OK : %d\n", c);
  }
  Serial.println("Done!");
```

The result :
All of color16bits, both before and after colors are the same color.

Great work!

**Comment by kisvegabor** [comment]
> The new algorithm from @paulpv
for transforming color16bits to color32bits
and transform backward by lv_color_make(...).

Great news! 
As the 16->32 and 32->16 bit conversations are the most common I suggest merging this fix now and
 fix the other conversation in new PRs if you have time to deal with them.

File: src/lv_misc/lv_color.h [hunk 1]
```diff
@@ -305,9 +305,14 @@ static inline uint32_t lv_color_to32(lv_color_t color)
 #elif LV_COLOR_DEPTH == 16
 #if LV_COLOR_16_SWAP == 0
     lv_color32_t ret;
-    ret.ch.red   = color.ch.red * 8;   /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
-    ret.ch.green = color.ch.green * 4; /*(2^8 - 1)/(2^6 - 1) = 255/63 = 4*/
-    ret.ch.blue  = color.ch.blue * 8;  /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
+    /**
+     * Per https://docs.google.com/spreadsheets/d/1PppX8FJpddauAPasHwlNgIPGIuPGPNvRbhilIQ8w-7g/edit#gid=0
```
Comment (embeddedt) [inline src/lv_misc/lv_color.h]: Not a big deal, but this spreadsheet doesn't seem to be publicly viewable.

I suppose that either that should be changed or this comment should be removed.

Comment (paulpv) [reply src/lv_misc/lv_color.h]: Will move abbreviated logic in to code comment...

Comment (paulpv) [reply src/lv_misc/lv_color.h]: BTW, I thought I had made the spreadsheet public, but I guess it didn't take.
I made it public and confirmed it is.
I'll still abbreviate the logic in a code comment...

**Comment by paulpv** [comment]
I found "smaller" math that has the same result.
This makes sense.
The trick is more precisely to find the first `>>` value that cancels the `+` value when converting `0` that has and a min result of 0 and max result of 255.
I am still working on a concise comment to put in the code...

**Comment by paulpv** [comment]
Feedback/suggestions/edit of the not so concise comment are welcome

**Comment by kisvegabor** [comment]
@paulpv  Sorry for the concise comment but I can't say anything than looks good to me. :slightly_smiling_face:

**Comment by paulpv** [comment]
Thanks. Obviously, feel free to merge at your discretion.

**Comment by kisvegabor** [comment]
Merged, thank you!

**Pull request merged**